File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/legacy/rest/api/CryptoApi.java
/*
 * Copyright 2014 OpenMarket Ltd
 * Copyright 2017 Vector Creations Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package im.vector.matrix.android.internal.legacy.rest.api;


import im.vector.matrix.android.internal.legacy.rest.model.pid.DeleteDeviceParams;
import im.vector.matrix.android.internal.legacy.rest.model.sync.DevicesListResponse;
import im.vector.matrix.android.internal.legacy.rest.model.crypto.KeyChangesResponse;
import im.vector.matrix.android.internal.legacy.rest.model.crypto.KeysClaimResponse;
import im.vector.matrix.android.internal.legacy.rest.model.crypto.KeysQueryResponse;
import im.vector.matrix.android.internal.legacy.rest.model.crypto.KeysUploadResponse;

import java.util.Map;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.HTTP;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface CryptoApi {

    /**
     * Upload device and/or one-time keys.
     * @param params the params.
     */
    @POST("keys/upload")
    Call<KeysUploadResponse> uploadKeys(@Body Map<String, Object> params);

    /**
     * Upload device and/or one-time keys.
     *
     * @param deviceId the deviceId
     * @param params   the params.
     */
    @POST("keys/upload/{deviceId}")
    Call<KeysUploadResponse> uploadKeys(@Path("deviceId") String deviceId, @Body Map<String, Object> params);

    /**
     * Download device keys.
     * @param params the params.
     */
    @POST("keys/query")
    Call<KeysQueryResponse> downloadKeysForUsers(@Body Map<String, Object> params);

    /**
     * Claim one-time keys.
     * @param params the params.
     */
    @POST("keys/claim")
    Call<KeysClaimResponse> claimOneTimeKeysForUsersDevices(@Body Map<String, Object> params);

    /**
     * Send an event to a specific list of devices
     *
     * @param eventType     the type of event to send
     * @param transactionId the random path item
     * @param params        the params
     */
    @PUT("sendToDevice/{eventType}/{random}")
    Call<Void> sendToDevice(@Path("eventType") String eventType, @Path("random") String transactionId, @Body Map<String, Object> params);

    /**
     * Get the devices list
     */
    @GET("devices")
    Call<DevicesListResponse> getDevices();

    /**
     * Delete a device.
     *
     * @param deviceId the device id
     * @param params   the deletion parameters
     */
    @HTTP(path = "devices/{device_id}", method = "DELETE", hasBody = true)
    Call<Void> deleteDevice(@Path("device_id") String deviceId, @Body DeleteDeviceParams params);

    /**
     * Update the device information.
     *
     * @param deviceId the device id
     * @param params   the params
     */
    @PUT("devices/{device_id}")
    Call<Void> updateDeviceInfo(@Path("device_id") String deviceId, @Body Map<String, String> params);

    /**
     * Get the update devices list from two sync token.
     *
     * @param oldToken the start token.
     * @param newToken the up-to token.
     */
    @GET("keys/changes")
    Call<KeyChangesResponse> getKeyChanges(@Query("from") String oldToken, @Query("to") String newToken);
}

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/api/CryptoApi.kt
/*
 * Copyright 2014 OpenMarket Ltd
 * Copyright 2017 Vector Creations Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package im.vector.matrix.android.internal.crypto.api

import im.vector.matrix.android.internal.crypto.model.rest.*
import im.vector.matrix.android.internal.network.NetworkConstants
import retrofit2.Call
import retrofit2.http.*

internal interface CryptoApi {

    /**
     * Get the devices list
     * Doc: https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-devices
     */
    @GET(NetworkConstants.URI_API_PREFIX_PATH_R0 + "devices")
    fun getDevices(): Call<DevicesListResponse>

    /**
     * Get the device info by id
     * Doc: https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-devices-deviceid
     */
    @GET(NetworkConstants.URI_API_PREFIX_PATH_R0 + "devices/{deviceId}")
    fun getDeviceInfo(@Path("deviceId") deviceId: String): Call<DeviceInfo>

    /**
     * Upload device and/or one-time keys.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-keys-upload
     *
     * @param params the params.
     */
    @POST(NetworkConstants.URI_API_PREFIX_PATH_R0 + "keys/upload")
    fun uploadKeys(@Body body: KeysUploadBody): Call<KeysUploadResponse>

    /**
     * Upload device and/or one-time keys.
     * Doc: not documented
     *
     * @param deviceId the deviceId
     * @param params   the params.
     */
    @POST(NetworkConstants.URI_API_PREFIX_PATH_R0 + "keys/upload/{deviceId}")
    fun uploadKeys(@Path("deviceId") deviceId: String, @Body body: KeysUploadBody): Call<KeysUploadResponse>

    /**
     * Download device keys.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-keys-query
     *
     * @param params the params.
     */
    @POST(NetworkConstants.URI_API_PREFIX_PATH_R0 + "keys/query")
    fun downloadKeysForUsers(@Body params: KeysQueryBody): Call<KeysQueryResponse>

    /**
     * Claim one-time keys.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#post-matrix-client-r0-keys-claim
     *
     * @param params the params.
     */
    @POST(NetworkConstants.URI_API_PREFIX_PATH_R0 + "keys/claim")
    fun claimOneTimeKeysForUsersDevices(@Body body: KeysClaimBody): Call<KeysClaimResponse>

    /**
     * Send an event to a specific list of devices
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#put-matrix-client-r0-sendtodevice-eventtype-txnid
     *
     * @param eventType     the type of event to send
     * @param transactionId the transaction ID for this event
     * @param body          the body
     */
    @PUT(NetworkConstants.URI_API_PREFIX_PATH_R0 + "sendToDevice/{eventType}/{txnId}")
    fun sendToDevice(@Path("eventType") eventType: String, @Path("txnId") transactionId: String, @Body body: SendToDeviceBody): Call<Unit>

    /**
     * Delete a device.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#delete-matrix-client-r0-devices-deviceid
     *
     * @param deviceId the device id
     * @param params   the deletion parameters
     */
    @HTTP(path = NetworkConstants.URI_API_PREFIX_PATH_R0 + "devices/{device_id}", method = "DELETE", hasBody = true)
    fun deleteDevice(@Path("device_id") deviceId: String, @Body params: DeleteDeviceParams): Call<Unit>

    /**
     * Update the device information.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#put-matrix-client-r0-devices-deviceid
     *
     * @param deviceId the device id
     * @param params   the params
     */
    @PUT(NetworkConstants.URI_API_PREFIX_PATH_R0 + "devices/{device_id}")
    fun updateDeviceInfo(@Path("device_id") deviceId: String, @Body params: UpdateDeviceInfoBody): Call<Unit>

    /**
     * Get the update devices list from two sync token.
     * Doc: https://matrix.org/docs/spec/client_server/r0.4.0.html#get-matrix-client-r0-keys-changes
     *
     * @param oldToken the start token.
     * @param newToken the up-to token.
     */
    @GET(NetworkConstants.URI_API_PREFIX_PATH_R0 + "keys/changes")
    fun getKeyChanges(@Query("from") oldToken: String, @Query("to") newToken: String): Call<KeyChangesResponse>
}
